broadinstitute/pearl
# JN-190 in-progress survey saving

This adds auto-saving every 3 seconds for surveys. Per direction from Erin and Jeff, "submitting" a survey now doesn't mean anything except marking a task and response as "complete".  Answers are saved and made available across the application as soon as they are entered.  And "completed" surveys can now be edited.  We use DataChangeRecords to create an audit trail of changes. 

 The autosave is on a timer, rather than on an answer value changing, both because surveyJS doesn't make a simple onchange API (there are different onchange handlers for every question type), and also because then we don't have to worry about too many save requests coming too quickly if a user quickly moves through a form. 

When the survey is complete, we resubmit the entire form, and then the server takes care of updating any answers that were missed by the autosave.  This is to ensure we get everything, just in case some weird timing issue or diff edge case means we missed some answers along the way.   This does mean it's possible for answers to clobber answers submitted in the interim, (if the user starts a survey, then adds some answers in another browser, and then resumed the survey in this original browser without reloading), but we prioritize making sure we get everything (and simpler implementation) over a true shared-editing sync experience.   Because even if an answer gets clobbered, it will leave an audit trail.  We might upgrade to sync-editing when our survey infrastructure is more mature and/or we implement admin survey editing.  

A later UX change will be to add an explicit "save for later" button to the UI

TO TEST:
1. restart apiParticipantApp, and apiAdminApp, and repopulate OurHealth 
2. sign in as consented@test.com
3. start a survey, fill out some answers
4. reload the page -- confirm you didn't lose too much data (< 3 seconds worth, although I think that an answer has to lose focus to register as a change in SurveyJS at all) 
5. close your browser
6. reopen, and go back to the hub signed in as yourself.  You should see the task as "In progress".  Going back should show you the answers you've already filled out
7.  Go to the admin tool, `https://localhost:3000/ourhealth/studies/ourheart/env/sandbox/participants/OHSENT/` -- you should see a task in progress.  Clicking on the survey you edited should show the answers.
